compiler/packages/babel-plugin-react-compiler/src/Inference/DropManualMemoization.ts

```diff
@@ -476,3 +484,46 @@ export function dropManualMemoization(func: HIRFunction): void {
     }
   }
 }
+
+function findOptionalPlaces(fn: HIRFunction): Set<IdentifierId> {
+  const optionals = new Set<IdentifierId>();
+  for (const [, block] of fn.body.blocks) {
+    if (block.terminal.kind === 'optional') {
+      const optionalTerminal = block.terminal;
+      let testBlock = fn.body.blocks.get(block.terminal.test)!;
+      loop: while (true) {
+        const terminal = testBlock.terminal;
+        switch (terminal.kind) {
+          case 'branch': {
+            if (terminal.fallthrough === optionalTerminal.fallthrough) {
+              // found it
+              const consequent = fn.body.blocks.get(terminal.consequent)!;
+              const last = consequent.instructions.at(-1);
+              if (last !== undefined && last.value.kind === 'StoreLocal') {
+                optionals.add(last.value.value.identifier.id);
+              }
```

Comment:
This feels slightly weird since DropManualMemo runs before we enter ssa, so both consequent (the optional read) and alternate (`undefined`) write to the same identifier ID. There is also no phi join, so reads to the OptionalExpression rvalue is also guaranteed to use the same identifier ID (and this works)

(I'm guessing that passes after SSA will need to track phis as well -- reading the rest of the stack now)

Reply:
Yeah I looked at using the phi to figure out which values is optional. First, as you noted we aren't in SSA so we don't have phis yet. Second, the phis end up in weird places for nested optionals, because with nested optionals the alternate actually goes to the outermost fallthrough. So we just look at where the consequent assigns to in order to figure out what the optional value was.